### PR TITLE
ramips: add support for Arcadyan WG630223

### PR DESCRIPTION
**Pull request by system1357**
Arcadyan WG630223 is a wireless AX1800 (WiFi 6) mesh router. It uses
the same case design as MTS WG430223.

Device specification
--------------------
SoC Type: MediaTek MT7621AT
RAM: 256 MiB
Flash: 128 MiB (Toshiba TC58BVG0S3HTA00 SLC SPI NAND)
Wireless 2.4 GHz (MT7915DA): b/g/n/ax, 2x2
Wireless 5 GHz (MT7915DN): a/n/ac/ax, 2x2
Ethernet: 3xGbE (WAN, LAN1, LAN2)
USB ports: No
Button: 1 (Reset/WPS)
LEDs: 3 (Red, Blue, Green)
Power: 12 VDC, 1 A
Connector type: Barrel
Bootloader: U-Boot (Version: 2018.09_TC-0.2)
Also called: CHT Wi-Fi Home-NFP (Wi-Fi 6_2T2R)

Installation(ssh)
------------
1. Login to the router web interface (username: cht, password: 23tc +
    last 4 digits of label mac address)
2. Navigate to System -> Remote Management -> enable ssh for LAN
3. Connect to router using scp (credentials same as web interface)
    to upload firmware, default folder is /tmp/cht_home
4. Connect to the router using ssh (credentials same as web interface)
5. Flash OpenWrt image:
```
ls -l (this is to find out file length)
mtd_debug erase /dev/mtd4 0 25165824
mtd_debug write /dev/mtd4 0 "file length" "file name"
```
9. Reboot into OpenWrt

Installation(TFTP / uboot console)
------------

1. Prepare a TFTP server (tested with tftpd32)
2. Connect TX and GND pins first (UART provision is near the reset button)
    Pin arrangement from arrow sign to J4 silkscreen is PWR TX RX GND
    Do not ever connect power pin, don't connect RX just yet
3. Serial console baud rate 115200 8N1, connect RX once there's output.
    Press any key to stop auto boot, beware that the window is tight
4. (optional) Set bootdelay to longer period for easier uboot access
`env set bootdelay 10`
5. Flash OpenWrt image:
```
run 'mtkautoboot'
> Press arrow down key immediately to stop autoboot
> Select option "Upgrade firmware"
> Select upgrade method "tftp", xmodem is slow but works as well
> Fill in network data and update file name, press enter
```
6. Router auto boots into OpenWrt

Stock firmware image format
---------------------------
The same as MTS WG630223 but with another trx magic "HDR0"

Others
---------------------------
This branch also contains the necessary fix for decrypting the
board_config partition using different keys

### CODE REVIEW

File: target/linux/ramips/mt7621/base-files/lib/upgrade/platform.sh
Comment (csharper2005) [inline target/linux/ramips/mt7621/base-files/lib/upgrade/platform.sh:56]: This will brake other devices. You need only one line here.
`arcadyan,wg630223|\`

File: target/linux/ramips/mt7621/base-files/etc/board.d/02_network
Comment (csharper2005) [inline target/linux/ramips/mt7621/base-files/etc/board.d/02_network:14]: This also brakes other devices.

File: target/linux/ramips/dts/mt7621_arcadyan_wg630223.dts
Comment (csharper2005) [inline target/linux/ramips/dts/mt7621_arcadyan_wg630223.dts:169]: Do we need to remove this block after https://github.com/openwrt/openwrt/pull/10238?

File: target/linux/ramips/dts/mt7621_arcadyan_wg630223.dts
Comment (csharper2005) [inline target/linux/ramips/dts/mt7621_arcadyan_wg630223.dts:49]: Hmm. Really 476?

Comment (system1357) [reply target/linux/ramips/dts/mt7621_arcadyan_wg630223.dts]: Ooops, that's indeed wrong

File: target/linux/ramips/dts/mt7621_arcadyan_wg630223.dts
Comment (system1357) [reply target/linux/ramips/dts/mt7621_arcadyan_wg630223.dts]: I've removed it for now

**Comment by jospezial** [comment]
If it has ssh, does the stock firmware have sftp server or client? sshfs?
A pipe over ssh would also be possible.
Then nobody needs to setup a http server.

**Comment by system1357** [comment]
> If it has ssh, does the stock firmware have sftp server or client? sshfs? A pipe over ssh would also be possible. Then nobody needs to setup a http server.

Fixed PR info, added tftp/uboot and ssh flash instructions

**Comment by csharper2005** [comment]
@system1357, In general looks good to me. I would suggest:
1. Make "u-boot-env" writeable;
2. Add MAC address examples to the commit message.

And I would move the changes to base-files and already supported devices in a separate commit, e.g. "base-files: Allow custom encryption key for Arcadyan".

Perhaps maintainers have a different opinion on this matter. Let's wait for their review.

**Review by csharper2005 — APPROVED** [review]

**Comment by jospezial** [comment]
> * don't connect RX just yet
> 
> * Serial console baud rate 115200 8N1, connect RX once there's output.

Why? What happens if it is connected before?

**Comment by system1357** [comment]
> Why? What happens if it is connected before?

There will be no serial output. But if the TX pin is connected after the serial console is active(started printing something), everything works fine throughout the whole powerup session
(yes, even after a reboot command)

I've tested with a few different serial console program and adapters with different chips, all seems to show the same behavior

**Comment by system1357** [comment]
> @system1357 Please take a look at this PR #10451

I believe this device has a DFS antenna, which the stock firmware also utilize for scanning
(The one at lower right)
6411D2F1-7457-4F64-826E-A462FDC42F17

File: target/linux/ramips/dts/mt7621_arcadyan_wg630223.dts
Comment (DragonBluep) [inline target/linux/ramips/dts/mt7621_arcadyan_wg630223.dts:64]: It's better to add `mediatek,bmt-remap-range`. NMBM should exclude ubi partition. refer to [actiontec web7200](https://git.openwrt.org/?p=openwrt/staging/nbd.git;a=blob;f=target/linux/ramips/dts/mt7621_actiontec_web7200.dts)

File: target/linux/ramips/dts/mt7621_arcadyan_wg630223.dts
Comment (DragonBluep) [inline target/linux/ramips/dts/mt7621_arcadyan_wg630223.dts:123]: It is recommended to add a new line above.

File: target/linux/ramips/image/mt7621.mk
Comment (DragonBluep) [inline target/linux/ramips/image/mt7621.mk:176]: The new u-boot 2018.09 should no longer need the lzma loader.

File: target/linux/ramips/image/mt7621.mk
Comment (DragonBluep) [inline target/linux/ramips/image/mt7621.mk:185]: u-boot 2018.09 supports fit image. Maybe you can try it if you have time. 0f068e7c4a83bcbf20c4e52a5f8a3f1fe2af2246 13308161788c98ae6cd48c22b13339fdb8c77130

File: target/linux/ramips/dts/mt7621_arcadyan_wg630223.dts
Comment (system1357) [reply target/linux/ramips/dts/mt7621_arcadyan_wg630223.dts:64]: May I ask:
If the second and third firmware partition uses squashfs/jffs2 layout, shoud I include them in the remap range?

File: target/linux/ramips/image/mt7621.mk
Comment (system1357) [reply target/linux/ramips/image/mt7621.mk:185]: I'm actually running a FIT image on my own device, will push fixes later

File: target/linux/ramips/dts/mt7621_arcadyan_wg630223.dts
Comment (DragonBluep) [reply target/linux/ramips/dts/mt7621_arcadyan_wg630223.dts:64]: yes, it should be `mediatek,bmt-remap-range = <0x0000000 0x05c0000>, <0x1980000 0x3800000>;`.  Maybe we need to use u-boot to restore the original firmware.

File: target/linux/ramips/image/mt7621.mk
Comment (DragonBluep) [inline target/linux/ramips/image/mt7621.mk:201]: Please replace these nand common definitions with `$(Device/nand)`.
ref: https://github.com/openwrt/openwrt/commit/b09a83808a3a4f6ff2ff0cf1f12fdc4193357896

**Pull request closed without merging**
